src/xdsl/ir.py

```diff
@@ -368,18 +368,35 @@ def verify(self, verify_nested_ops: bool = True) -> None:
     def verify_(self) -> None:
         pass
 
-    def clone_without_regions(self: OperationType) -> OperationType:
+    def clone_without_regions(self: OperationType,
+                              mapper=None) -> OperationType:
         """Clone an operation, with empty regions instead."""
-        operands = self.operands
+        if mapper is None:
+            mapper = {}
+        operands = [(mapper[operand] if operand in mapper else operand)
+                    for operand in self.operands]
         result_types = [res.typ for res in self.results]
         attributes = self.attributes.copy()
-        successors = self.successors.copy()
+        successors = [(mapper[successor] if successor in mapper else successor)
+                      for successor in self.successors]
         regions = [Region() for _ in self.regions]
-        return self.create(operands=operands,
-                           result_types=result_types,
-                           attributes=attributes,
-                           successors=successors,
-                           regions=regions)
+        clonedOp = self.create(operands=operands,
+                               result_types=result_types,
+                               attributes=attributes,
+                               successors=successors,
+                               regions=regions)
+        for idx, result in enumerate(clonedOp.results):
+            mapper[self.results[idx]] = result
```

Comment:
I find that a bit ugly to have both the mapper being used for operands, successors, and also being updated for the results.
Since we have anyway the previous operation, can we remove this update, and let the user create their own mapper outside of this function?

Reply:
The use case where this is required is cloning a region with lots of operations, which also refer to another and possibly again have nested regions. 
consider the following example:
```
module() {
  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
  %1 : !i32 = arith.constant() ["value" = 2 : !i32]
  %2 : !i32 = arith.addi(%0 : !i32, %1 : !i32)
  std.return(%2 : !i32)
}
```
the call to `module.clone()` passes a mapper to all the operations in the region, which the two constants use to register the mapping of their old result to the result of the cloned operations.
addi finds the mapping of results for the two constants and uses the updated values of the mapper as operands. 

So it is necessary for an operation being cloned in context of a region to add its results to the map and to look up possible changes in the operands in the map. Same goes the successors of an op. The mapping of old Block to new Block is added by the Region being cloned. 

We could divide the `mapper` into a `valueMapper` for operands and results and a `blockMapper` for successors if you like that better. 

Another thing: If this function does not get passed a mapper we can assume the op is being cloned in isolation. So this could also look like:
```
    def clone_without_regions(self: OperationType,
                              mapper=None) -> OperationType:
        if mapper is None:
            # branch old version
            operands = self.operands
            result_types = [res.typ for res in self.results]
            attributes = self.attributes.copy()
            successors = self.successors.copy()
            regions = [Region() for _ in self.regions]
            return self.create(operands=operands,
                           result_types=result_types,
                           attributes=attributes,
                           successors=successors,
                           regions=regions)
        else:
            # branch PR version, what is now starting at line 376
```

but I did not like the duplication in it. Maybe you have an idea for something more elegant?

Reply:
Oh yes, thanks for the explanation, I understand better the problem now!
I think I would prefer a value_mapper and a block_mapper, to make the type information a bit more explicit!

Otherwise, I don't think we need the version you proposed when mapper is None, since it would essentially have the same result as the version you have in the PR? I think the only difference should be performances right? If that is the case, then I guess it's better to not have this code duplication, since we do not really care that much about performances right now, and more about correctness/readability!

Reply:
I committed the changes we discussed and will now convert from a draft PR to a PR.

Also note: I added the feature to specify an `insertion_index` when cloning from one region into another.